src/Observability/Runtime/Common/ExportFormatter.cs

```diff
@@ -148,6 +161,52 @@ public static string FormatLogData(IDictionary<string, object?> data)
             return resource.Attributes.FirstOrDefault(a => a.Key == "service.version").Value?.ToString();
         }
 
+        private static OtlpSpan BuildOtlpSpanWithTruncation(Activity activity, Action<string>? logInformation = null)
+        {
+            var span = BuildOtlpSpan(activity);
+
+            // Check initial size
+            if (Encoding.UTF8.GetByteCount(ExportFormatter.SerializePayload(span)) <= ExportFormatter.MaxSpanSizeBytes)
+                return span;
+
+            // Gather key sizes
+            var keySizes = new List<(string Key, int Size, string? Value)>();
+            foreach (var key in ExportFormatter.LargePayloadAttributeKeys)
+            {
+                if (span.Attributes != null && span.Attributes.TryGetValue(key, out var valueObj))
+                {
+                    var value = valueObj as string;
+                    int size = !string.IsNullOrEmpty(value) ? Encoding.UTF8.GetByteCount(value) : 0;
+                    keySizes.Add((key, size, value));
+                    logInformation?.Invoke($"Activity '{activity.DisplayName}': Key '{key}' size = {size / 1024} KB.");
+                }
+            }
+
+            // Sort keys by size descending
+            var sorted = keySizes
+                .Where(k => !string.IsNullOrEmpty(k.Value) && k.Size > 0)
+                .OrderByDescending(k => k.Size)
+                .ToList();
+
+            foreach (var (key, size, _) in sorted)
+            {
+                if (span.Attributes != null)
```

Comment:
nit: should we move this condition to line 167 so if `span.Attributes` is null we return early the span unmodified?